src/computational_graph/graph.jl

```diff
@@ -255,124 +282,212 @@ end
-- `external`  index of external vertices. They are the actual external quantum operators, not the ghost operators.
+- `subgraphs::Vector{Graph{F,W}}` all subgraphs of the diagram. All external operators of subgraphs constitute all operators of the new diagram.
+- `topology::Vector{Vector{Int}}` topology of the diagram. Each Vector{Int} stores operators' index connected with each other (as a propagator). 
+- `perm_noleg::Union{Vector{Int},Nothing}=nothing` permutation of all the nonleg external operators. By default, setting nothing means to use the default order from subgraphs.
 - `factor::F`  overall scalar multiplicative factor for this diagram (e.g., permutation sign)
 - `weight`  weight of the diagram
 - `name`  name of the diagram
 - `type`  type of the diagram
 
 # Example:
 ```julia-repl
-julia> g = feynman_diagram([𝑓⁺(1)𝑓⁻(2)𝜙(3), 𝑓⁺(4)𝑓⁻(5)𝜙(6)], [[5, 1], [2, 4], [3, 6]])
-4:f⁺(1)f⁻(2)ϕ(3)|f⁺(4)f⁻(5)ϕ(6)=0.0=-1.0Ⓧ (1,2,3)
+julia> V = [𝑓⁺(1)𝑓⁻(2)𝜙(3), 𝑓⁺(4)𝑓⁻(5)𝜙(6), 𝑓⁺(7)𝑓⁻(8)𝜙(9)];
+julia> g = feynman_diagram(interaction.(V), [[1, 5], [3, 9], [4, 8]], perm_noleg=[3, 1, 2])
+7:f⁺(1)f⁻(2)ϕ(3)|f⁺(4)f⁻(5)ϕ(6)|f⁺(7)f⁻(8)ϕ(9)=0.0=Ⓧ (1,2,3,4,5,6)
 
 julia> g.subgraphs
-3-element Vector{Graph{Float64, Float64}}:
-1:f⁻(5)|f⁺(1)=0.0
-2:f⁻(2)|f⁺(4)=0.0
-3:ϕ(3)|ϕ(6)=0.0
+6-element Vector{Graph{Float64, Float64}}:
+ 1:f⁺(1)f⁻(2)ϕ(3)=0.0
+ 2:f⁺(4)f⁻(5)ϕ(6)=0.0
+ 3:f⁺(7)f⁻(8)ϕ(9)=0.0
+ 4:f⁺(1)|f⁻(5)⋅-1.0=0.0
+ 5:ϕ(3)|ϕ(9)=0.0
+ 6:f⁺(4)|f⁻(8)⋅-1.0=0.0
 ```
 """
-function feynman_diagram(vertices::AbstractVector,
-    topology::Vector{Vector{Int}};
-    external=[], factor=one(_dtype.factor),
-    weight=zero(_dtype.weight),
-    name="", type=:generic)
-
-    # convert all Graphs to vertices by grouping all its external legs as a vertex
-    converted_vertices = [_extract_vertex(typeof(v), v) for v in vertices]
-    # collect graphs in vertices as subdiagrams
-    subdiagrams = [v for v in vertices if isa(v, Graph)]
-
-    operators = [o for v in converted_vertices for o in v.operators]
-    permutation = collect(Iterators.flatten(topology))
-    ind_ops = collect(eachindex(operators))
-
-
-    @assert length(unique(permutation)) == length(permutation) # no repeated index
-    @assert length(unique(external)) == length(external) # no repeated index
-    # @assert Set(permutation) == Set(ind_ops) # permutation must exhaust all operators
-    ind_ghost = filter(p -> isghost(operators[p]), ind_ops)
-    @assert all(ind_ghost .<= length(external)) # external real/fake legs must be placed at the beginning of vertices.
-
-    ind_fakeleg = Int[]
-    subgraphs = Graph[]
-    for connection in topology
-        if isempty(intersect(connection, ind_ghost))
-            push!(subgraphs, propagator(operators[connection]))
+function feynman_diagram(subgraphs::Vector{Graph{F,W}}, topology::Vector{Vector{Int}}; perm_noleg::Union{Vector{Int},Nothing}=nothing,
```

Comment:
maybe it is better to define perm_noleg as the third non-keyword optional argument. This name is too difficult to remeber